test/e2e/models/CodeServer.ts

```diff
@@ -0,0 +1,104 @@
+import { Page } from "playwright"
+import { CODE_SERVER_ADDRESS } from "../../utils/constants"
+// This is a Page Object Model
+// We use these to simplify e2e test authoring
+// See Playwright docs: https://playwright.dev/docs/pom/
+export class CodeServer {
+  page: Page
+
+  constructor(page: Page) {
+    this.page = page
+  }
+
+  /**
+   * Navigates to CODE_SERVER_ADDRESS
+   */
+  async navigate() {
+    await this.page.goto(CODE_SERVER_ADDRESS, { waitUntil: "networkidle" })
+  }
+
+  /**
+   * Checks if the editor is visible
+   * and reloads until it is
+   */
+  async reloadUntilEditorIsVisible() {
+    const editorIsVisible = await this.isEditorVisible()
+    let reloadCount = 0
+
+    // Occassionally code-server timeouts in Firefox
+    // we're not sure why
+    // but usually a reload or two fixes it
+    // TODO@jsjoeio @oxy look into Firefox reconnection/timeout issues
+    while (!editorIsVisible) {
+      reloadCount += 1
+      if (await this.isEditorVisible()) {
+        console.log(`    Editor became visible after ${reloadCount} reloads`)
+        break
+      }
+      // When a reload happens, we want to wait for all resources to be
+      // loaded completely. Hence why we use that instead of DOMContentLoaded
+      // Read more: https://thisthat.dev/dom-content-loaded-vs-load/
+      await this.page.reload({ waitUntil: "load" })
```

Comment:
The fact that we have to do this is probably a bug — either in code-server or something with Firefox. Probably us though.

Reply:
I noticed, after introducing this logic, the flakiness for tests went completely away. Most times when tests were failing before, it was because code-server didn't load.